contracts/exchangeIssuance/FlashMintWrapped.sol

```diff
@@ -17,220 +17,1087 @@
-    {
-        return 0;
+  {
+    // safe approves are done right in the dexAdapter library
+    // swaps are skipped there too if inputToken == outputToken (depending on path)
+    return dexAdapter.swapTokensForExactTokens(_amount, _maxAmountIn, _swapDexData);
+  }
+
+  /**
+   * Swaps exact _amount of _inputToken to _outputToken through _swapDexData
+   *
+   * @param _inputToken           Input token that will be sold
+   * @param _outputToken          Output token that will be bought
+   * @param _amount               Amount that will be sold
+   * @param _swapDexData          DEXAdapter.SwapData with path, fees, etc. for inputToken -> outputToken swap
+   *
+   * @return amount of received _outputToken
+   */
+  function _swapFromExact(
+    IERC20 _inputToken,
+    IERC20 _outputToken,
+    uint256 _amount,
+    DEXAdapter.SwapData calldata _swapDexData
+  )
+    internal
+    isValidPath(_swapDexData.path, address(_inputToken), address(_outputToken))
+    returns (uint256)
+  {
+    // safe approves are done right in the dexAdapter library
+    return
+      dexAdapter.swapExactTokensForTokens(
+        _amount,
+        // _minAmountOut is 0 here since we don't know what to check against because for wrapped components
+        // we only have the required amounts for the wrapped component, but not for the underlying we swap to here
+        // This is covered indirectly in later checks though
+        // e.g. directly through the issue call (not enough _outputToken -> wrappedComponent -> issue will fail)
+        0,
+        _swapDexData
+      );
+  }
+
+  /**
+   * Wraps _wrapAmount of _underlyingToken to _wrappedComponent component
+   *
+   * @param _invokeWrapData             invoke wrap data for call target, call value and bytes calldata
+   * @param _wrapAmount                 amount of _underlyingToken to wrap
+   * @param _underlyingToken            underlying (unwrapped) token to wrap from (e.g. DAI)
+   * @param _wrappedToken               wrapped token to wrap to
+   */
+  function _wrapComponent(
+    ComponentInvokeWrapData calldata _invokeWrapData,
+    uint256 _wrapAmount,
+    address _underlyingToken,
+    address _wrappedToken
+  ) internal {
+    require(_invokeWrapData.callTarget != address(0) && _invokeWrapData.callData[0] != 0,"FlashMint: WRAP_DATA_MISSING");
+
+    require(_invokeWrapData.callTarget == _wrappedToken || wrapTargetsAllowList[_invokeWrapData.callTarget],"FlashMint: WRAP_CALLTARGET_NOT_ALLOWED");
+
+    // approve _underlyingToken token transfer from this to _wrapCallTarget
+    DEXAdapter._safeApprove(IERC20(_underlyingToken), _invokeWrapData.callTarget, _wrapAmount);
+    // invoke wrap function call. we can't check any response value because the implementation might be different 
+    // between wrapCallTargets... e.g. compoundV2 would return uint256 with value 0 if successful
+    _invokeWrapData.callTarget.functionCallWithValue(_invokeWrapData.callData, _invokeWrapData.value);
```

Comment:
This looks dangerous to me. `_invokeWrapData.callData` is provided by the user right ? So couldn't he encode a transfer / approve call here to potentially drain leftover funds from other users ?

Reply:
Some potential "attack vectors". (haven't really tested any of those or even thought them through to the end so feel free to challenge).

1. User issues 0 amount of set token and instead of the wrapping encodes an `approve` transaction. Afterwards he has the power to spend this contracts wrapped tokens.
2. When there is some underlying token left in the contract the user just wraps that and uses it to issue set token.

Reply:
I had a discussion regarding this with Richard. I had it initially with direct calls to our wrap adapters as I think you have seen. 
callData is provided by the user, yes. But the callTarget is allowListed or a component of set tokens see line 980. AllowList is in control of the contract owner (IndexCoop), SetToken components can also only be added by IndexCoop.

Reply:
Good points @ckoopmann, really appreciate your thinking about potential attack vectors here. 
Let's discuss in the other comment regarding if it's better to go back to wrap adapters...